plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php

```diff
@@ -115,11 +115,91 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
 				)
 			);
 
-		// phpcs:ignore WordPress.PHP.DontExtract.extract_extract
-		extract( $this->get_order_data( $order ) );
+		/**
+		 * Filter to customize the set of data that is used to render the receipt.
+		 * The formatted line items aren't included, use the woocommerce_printable_order_receipt_formatted_line_item
+		 * filter to customize those.
+		 *
+		 * See the value returned by the 'get_order_data' and 'get_woo_pay_data' methods for a reference of
+		 * the structure of the data.
+		 *
+		 * See the template file, Templates/order-receipt.php, for reference on how the data is used.
+		 *
+		 * @param array $data The original set of data.
+		 * @param WC_Order $order The order for which the receipt is being generated.
+		 * @returns array The updated set of data.
+		 *
+		 * @since 9.0.0
+		 */
+		$data = apply_filters( 'woocommerce_printable_order_receipt_data', $this->get_order_data( $order ), $order );
+
+		$formatted_line_items = array();
+		$row_index            = 0;
+		foreach ( $data['line_items'] as $line_item_data ) {
+			$quantity_data          = isset( $line_item_data['quantity'] ) ? " × ${line_item_data['quantity']}" : '';
+			$line_item_display_data = array(
+				'inner_html'    => "<td>${line_item_data['title']}$quantity_data</td><td>${line_item_data['amount']}</td>",
+				'tr_attributes' => array(),
+				'row_index'     => $row_index++,
+			);
```

Comment:
Use of `${var}` instead of `{$var}` is now deprecated, so we better update these now.